File: Backend/app/Services/AlertService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Enums\AlertSeverity;
use App\Enums\AlertStatus;
use App\Enums\EventType;
use App\Exceptions\AlertGenerationException;
use App\Models\Alert;
use App\Models\AlertRule;
use App\Models\Machine;
use App\Models\MachineCurrentStatus;
use Exception;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Validator;
use Illuminate\Validation\ValidationException;

/**
 * Class AlertService
 *
 * Handles alert generation, evaluation, acknowledgement, and resolution.
 * Evaluates machine health data against configured alert rules and
 * dispatches notifications for critical conditions.
 *
 * @package App\Services
 */
class AlertService
{
    /**
     * The notification service for dispatching alert notifications.
     *
     * @var NotificationService
     */
    private NotificationService $notificationService;

    /**
     * The audit log service for recording alert events.
     *
     * @var AuditLogService
     */
    private AuditLogService $auditLogService;

    /**
     * AlertService constructor.
     *
     * @param NotificationService $notificationService
     * @param AuditLogService     $auditLogService
     */
    public function __construct(NotificationService $notificationService, AuditLogService $auditLogService)
    {
        $this->notificationService = $notificationService;
        $this->auditLogService = $auditLogService;
    }

    /**
     * Evaluate all enabled alert rules against the machine's current status.
     *
     * For each rule that matches the current status values, an alert is created
     * and a notification is sent to the assigned user.
     *
     * @param  Machine              $machine
     * @param  MachineCurrentStatus $status
     * @return void
     *
     * @throws AlertGenerationException
     */
    public function evaluateMachineAlerts(Machine $machine, MachineCurrentStatus $status): void
    {
        try {
            $rules = AlertRule::where('company_id', $machine->company_id)
                ->where('is_enabled', true)
               ->get();

            foreach ($rules as $rule) {
                try {
                    if ($this->evaluateRule($rule, $status)) {
                        $this->createAlertFromRule($rule, $machine, $status);
                    }
                } catch (Exception $e) {
                    Log::warning('AlertService::evaluateMachineAlerts - Rule evaluation failed', [
                        'rule_id'    => $rule->id,
                        'machine_id' => $machine->id,
                        'error'      => $e->getMessage(),
                    ]);
                }
            }
        } catch (Exception $e) {
            Log::error('AlertService::evaluateMachineAlerts - Failed to evaluate alerts', [
                'machine_id' => $machine->id,
                'error'      => $e->getMessage(),
            ]);
            throw new AlertGenerationException(
                'Failed to evaluate alert rules for machine.',
                500,
                ['machine_id' => $machine->id]
            );
        }
    }

    /**
     * Acknowledge an alert.
     *
     * @param  int  $alertId
     * @param  int  $userId
     * @return Alert
     */
    public function acknowledgeAlert(int $alertId, int $userId): Alert
    {
        try {
            $alert = Alert::findOrFail($alertId);
            $oldValues = $alert->toArray();

            $alert->update([
                'status'           => AlertStatus::Acknowledged->value,
                'acknowledged_by'  => $userId,
                'acknowledged_at'  => now(),
            ]);

            $alert->refresh();

            $this->auditLogService->log(
                EventType::Update->value,
                'Alert acknowledged: ' . $alert->title,
                $oldValues,
                $alert->toArray(),
                null,
                $alert->machine
            );

            Log::info('Alert acknowledged', [
                'alert_id'   => $alert->id,
                'user_id'    => $userId,
                'machine_id' => $alert->machine_id,
            ]);

            return $alert;
        } catch (Exception $e) {
            Log::error('AlertService::acknowledgeAlert - Failed to acknowledge alert', [
                'alert_id' => $alertId,
                'user_id'  => $userId,
                'error'    => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Resolve an alert with an optional resolution note.
     *
     * @param  int         $alertId
     * @param  int         $userId
     * @param  string|null $resolution
     * @return Alert
     */
    public function resolveAlert(int $alertId, int $userId, ?string $resolution = null): Alert
    {
        try {
            $alert = Alert::findOrFail($alertId);
            $oldValues = $alert->toArray();

            $metadata = $alert->metadata ?? [];
            if ($resolution !== null) {
                $metadata['resolution'] = $resolution;
            }

            $alert->update([
                'status'      => AlertStatus::Resolved->value,
                'resolved_by' => $userId,
                'resolved_at' => now(),
                'metadata'    => $metadata,
            ]);

            $alert->refresh();

            $this->auditLogService->log(
                EventType::Update->value,
                'Alert resolved: ' . $alert->title . ($resolution ? ' - ' . $resolution : ''),
                $oldValues,
                $alert->toArray(),
                null,
                $alert->machine
            );

            Log::info('Alert resolved', [
                'alert_id'   => $alert->id,
                'user_id'    => $userId,
                'resolution' => $resolution,
            ]);

            return $alert;
        } catch (Exception $e) {
            Log::error('AlertService::resolveAlert - Failed to resolve alert', [
                'alert_id' => $alertId,
                'user_id'  => $userId,
                'error'    => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Retrieve alerts for a company, optionally filtered by severity and status.
     *
     * @param  int         $companyId
     * @param  string|null $severity
     * @param  string|null $status
     * @return Collection<int, Alert>
     */
    public function getCompanyAlerts(int $companyId, ?string $severity = null, ?string $status = null): Collection
    {
        try {
            $query = Alert::with(['machine', 'acknowledgedBy', 'resolvedBy'])
                ->where('company_id', $companyId);

            if ($severity !== null) {
                $query->where('severity', $severity);
            }

            if ($status !== null) {
                $query->where('status', $status);
            }

            return $query->orderBy('created_at', 'desc')->get();
        } catch (Exception $e) {
            Log::error('AlertService::getCompanyAlerts - Failed to retrieve company alerts', [
                'company_id' => $companyId,
                'severity'   => $severity,
                'status'     => $status,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Retrieve all alerts for a specific machine.
     *
     * @param  int  $machineId
     * @return Collection<int, Alert>
     */
    public function getMachineAlerts(int $machineId): Collection
    {
        try {
            return Alert::with(['acknowledgedBy', 'resolvedBy'])
                ->where('machine_id', $machineId)
                ->orderBy('created_at', 'desc')
                ->get();
        } catch (Exception $e) {
            Log::error('AlertService::getMachineAlerts - Failed to retrieve machine alerts', [
                'machine_id' => $machineId,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Retrieve all critical alerts for a company.
     *
     * @param  int  $companyId
     * @return Collection<int, Alert>
     */
    public function getCriticalAlerts(int $companyId): Collection
    {
        try {
            return Alert::with(['machine', 'acknowledgedBy', 'resolvedBy'])
                ->where('company_id', $companyId)
                ->where('severity', AlertSeverity::Critical->value)
                ->whereIn('status', [AlertStatus::Open->value, AlertStatus::Acknowledged->value])
                ->orderBy('created_at', 'desc')
                ->get();
        } catch (Exception $e) {
            Log::error('AlertService::getCriticalAlerts - Failed to retrieve critical alerts', [
                'company_id' => $companyId,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Evaluate a single rule against the current machine status.
     *
     * @param  AlertRule             $rule
     * @param  MachineCurrentStatus  $status
     * @return bool
     */
    private function evaluateRule(AlertRule $rule, MachineCurrentStatus $status): bool
    {
        $metricValue = $this->getMetricValue($rule->metric, $status);

        if ($metricValue === null) {
            return false;
        }

        $threshold = (float) ($rule->value ?? 0);

        return match ($rule->operator) {
            '>'  => $metricValue > $threshold,
            '>=' => $metricValue >= $threshold,
            '<'  => $metricValue < $threshold,
            '<=' => $metricValue <= $threshold,
            '==' => $metricValue == $threshold,
            '!=' => $metricValue != $threshold,
            default => false,
        };
    }

    /**
     * Extract the relevant metric value from the machine status.
     *
     * @param  string|null            $metric
     * @param  MachineCurrentStatus   $status
     * @return float|null
     */
    private function getMetricValue(?string $metric, MachineCurrentStatus $status): ?float
    {
        return match ($metric) {
            'cpu_percentage'  => $status->cpu_percentage !== null ? (float) $status->cpu_percentage : null,
            'cpu_temperature' => $status->cpu_temperature !== null ? (float) $status->cpu_temperature : null,
            'ram_percentage'  => $status->ram_percentage !== null ? (float) $status->ram_percentage : null,
            'disk_percentage' => $status->disk_percentage !== null ? (float) $status->disk_percentage : null,
            'battery_percentage' => $status->battery_percentage !== null ? (float) $status->battery_percentage : null,
            default => null,
        };
    }

    /**
     * Create an alert record from a triggered rule.
     *
     * @param  AlertRule             $rule
     * @param  Machine               $machine
     * @param  MachineCurrentStatus  $status
     * @return void
     */
    private function createAlertFromRule(AlertRule $rule, Machine $machine, MachineCurrentStatus $status): void
    {
        try {
            $severity = $rule->severity ?? AlertSeverity::Warning->value;

            $alert = Alert::create([
                'company_id'    => $machine->company_id,
                'machine_id'    => $machine->id,
                'alert_rule_id' => $rule->id,
                'severity'      => $severity,
                'title'         => $rule->name,
                'description'   => $this->buildAlertMessage($rule, $status),
                'metadata'      => [
                    'metric' => $rule->metric,
                    'operator' => $rule->operator,
                    'threshold' => $rule->value,
                    'current_value' => $this->getMetricValue($rule->metric, $status),
                ],
                'status'        => AlertStatus::Open->value,
            ]);

            $this->auditLogService->log(
                EventType::Create->value,
                'Alert generated: ' . $rule->name . ' for machine: ' . $machine->machine_uid,
                null,
                $alert->toArray(),
                null,
                $machine
            );

            $this->notificationService->sendAlertNotification($alert);

            // Send email notifications for critical and warning alerts
            if (in_array($severity, ['critical', 'warning'])) {
                $this->notificationService->sendEmailNotification($alert);
            }

            Log::info('Alert created from rule', [
                'alert_id'   => $alert->id,
                'rule_id'    => $rule->id,
                'machine_id' => $machine->id,
                'severity'   => $severity,
            ]);
        } catch (Exception $e) {
            Log::error('AlertService::createAlertFromRule - Failed to create alert', [
                'rule_id'    => $rule->id,
                'machine_id' => $machine->id,
                'error'      => $e->getMessage(),
            ]);
        }
    }

    /**
     * Build a human-readable alert message from the rule and current status.
     *
     * @param  AlertRule             $rule
     * @param  MachineCurrentStatus  $status
     * @return string
     */
    /**
     * Retrieve alert rules for a given company.
     *
     * @param  int  $companyId
     * @return Collection<int, AlertRule>
     */
    public function getAlertRules(int $companyId): Collection
    {
        try {
            return AlertRule::with(['company'])
                ->where('company_id', $companyId)
                ->orderBy('name', 'asc')
                ->get();
        } catch (Exception $e) {
            Log::error('AlertService::getAlertRules - Failed to retrieve alert rules', [
                'company_id' => $companyId,
                'error'      => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Update an alert rule with validated data.
     * Prevents changing the company_id through mass assignment.
     *
     * @param  int    $ruleId
     * @param  array  $data
     * @return AlertRule
     */
    public function updateAlertRule(int $ruleId, array $data): AlertRule
    {
        try {
            $rule = AlertRule::findOrFail($ruleId);

            // Prevent changing company_id via this method
            unset($data['company_id']);

            $rule->update($data);
            $rule->refresh();

            $this->auditLogService->log(
                EventType::Update->value,
                'Alert rule updated: ' . $rule->name,
                null,
                $rule->toArray(),
                null,
                null
            );

            Log::info('Alert rule updated', [
                'rule_id' => $rule->id,
                'data'    => $data,
            ]);

            return $rule;
        } catch (Exception $e) {
            Log::error('AlertService::updateAlertRule - Failed to update alert rule', [
                'rule_id' => $ruleId,
                'data'    => $data,
                'error'   => $e->getMessage(),
            ]);
            throw $e;
        }
    }

    /**
     * Build a human-readable alert message from the rule and current status.
     *
     * @param  AlertRule             $rule
     * @param  MachineCurrentStatus  $status
     * @return string
     */
    private function buildAlertMessage(AlertRule $rule, MachineCurrentStatus $status): string
    {
        $currentValue = $this->getMetricValue($rule->metric, $status);
        $metricLabel = str_replace('_', ' ', ucfirst($rule->metric ?? 'metric'));

        return sprintf(
            '%s is %s %s (current: %s)',
            $metricLabel,
            $rule->operator,
            $rule->value,
            $currentValue !== null ? (string) $currentValue : 'N/A'
        );
    }
}

File: DeskGuardBackend/Services/AlertService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using DeskGuardBackend.Data;
using DeskGuardBackend.Entities;
using DeskGuardBackend.Exceptions;
using DeskGuardBackend.Services.Interfaces;
using DeskGuardBackend.Enums;

namespace DeskGuardBackend.Services
{
    public class AlertService : IAlertService
    {
        private readonly DeskGuardDbContext _dbContext;
        private readonly INotificationService _notificationService;
        private readonly IAuditLogService _auditLogService;
        private readonly ILogger<AlertService> _logger;

        public AlertService(
            DeskGuardDbContext dbContext,
            INotificationService notificationService,
            IAuditLogService auditLogService,
            ILogger<AlertService> logger)
        {
            _dbContext = dbContext;
            _notificationService = notificationService;
            _auditLogService = auditLogService;
            _logger = logger;
        }

        public async Task EvaluateMachineAlertsAsync(Machine machine, MachineCurrentStatus status)
        {
            try
            {
                var rules = await _dbContext.AlertRules
                    .Where(r => r.CompanyId == machine.CompanyId && r.IsEnabled)
                    .ToListAsync();

                foreach (var rule in rules)
                {
                    try
                    {
                        if (EvaluateRule(rule, status))
                        {
                            await CreateAlertFromRuleAsync(rule, machine, status);
                        }
                    }
                    catch (Exception ex)
                    {
                        _logger.LogWarning(ex, "AlertService::EvaluateMachineAlertsAsync - Rule evaluation failed for rule ID {RuleId}, machine {MachineId}", rule.Id, machine.Id);
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::EvaluateMachineAlertsAsync - Failed to evaluate alerts for machine: {MachineId}", machine.Id);
                throw new AlertGenerationException("Failed to evaluate alert rules for machine.", 500);
            }
        }

        public async Task<Alert> AcknowledgeAlertAsync(long alertId, long userId)
        {
            try
            {
                var alert = await _dbContext.Alerts
                    .Include(a => a.Machine)
                    .FirstOrDefaultAsync(a => a.Id == alertId);

                if (alert == null)
                {
                    throw new KeyNotFoundException($"Alert not found: {alertId}");
                }

                alert.Status = AlertStatus.Acknowledged.ToString().ToLowerInvariant();
                alert.AcknowledgedBy = userId;
                alert.AcknowledgedAt = DateTime.UtcNow;
                await _dbContext.SaveChangesAsync();

                await _auditLogService.LogAsync(
                    EventType.Acknowledge.ToString(),
                    $"Alert acknowledged: {alert.Title}",
                    user: await _dbContext.Users.FindAsync(userId),
                    machine: alert.Machine
                );

                _logger.LogInformation("Alert {AlertId} acknowledged by user {UserId}", alertId, userId);
                return alert;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::AcknowledgeAlertAsync failed for alert ID: {AlertId}", alertId);
                throw;
            }
        }

        public async Task<Alert> ResolveAlertAsync(long alertId, long userId, string? resolution)
        {
            try
            {
                var alert = await _dbContext.Alerts
                    .Include(a => a.Machine)
                    .FirstOrDefaultAsync(a => a.Id == alertId);

                if (alert == null)
                {
                    throw new KeyNotFoundException($"Alert not found: {alertId}");
                }

                var resolutionMetadata = new Dictionary<string, string>();
                if (alert.Metadata != null)
                {
                    try
                    {
                        resolutionMetadata = JsonSerializer.Deserialize<Dictionary<string, string>>(alert.Metadata) ?? new Dictionary<string, string>();
                    }
                    catch { }
                }

                if (resolution != null)
                {
                    resolutionMetadata["resolution"] = resolution;
                }

                alert.Status = AlertStatus.Resolved.ToString().ToLowerInvariant();
                alert.ResolvedBy = userId;
                alert.ResolvedAt = DateTime.UtcNow;
                alert.Metadata = JsonSerializer.Serialize(resolutionMetadata);
                await _dbContext.SaveChangesAsync();

                await _auditLogService.LogAsync(
                    EventType.Resolve.ToString(),
                    $"Alert resolved: {alert.Title} {(resolution != null ? " - " + resolution : "")}",
                    user: await _dbContext.Users.FindAsync(userId),
                    machine: alert.Machine
                );

                _logger.LogInformation("Alert {AlertId} resolved by user {UserId}", alertId, userId);
                return alert;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::ResolveAlertAsync failed for alert ID: {AlertId}", alertId);
                throw;
            }
        }

        public async Task<PaginatedResponseDto<Alert>> GetCompanyAlertsAsync(
            long companyId, 
            string? severity, 
            string? status, 
            int page, 
            int perPage)
        {
            try
            {
                var query = _dbContext.Alerts
                    .Include(a => a.Machine)
                    .Include(a => a.Acknowledger)
                    .Include(a => a.Resolver)
                    .Where(a => a.CompanyId == companyId);

                if (!string.IsNullOrEmpty(severity))
                {
                    query = query.Where(a => a.Severity == severity.ToLowerInvariant());
                }

                if (!string.IsNullOrEmpty(status))
                {
                    query = query.Where(a => a.Status == status.ToLowerInvariant());
                }

                var total = await query.CountAsync();
                perPage = Math.Min(Math.Max(1, perPage), 100);
                var lastPage = (int)Math.Ceiling((double)total / perPage);
                page = Math.Min(Math.Max(1, page), Math.Max(1, lastPage));

                var items = await query
                    .OrderByDescending(a => a.CreatedAt)
                    .Skip((page - 1) * perPage)
                    .Take(perPage)
                    .ToListAsync();

                return new PaginatedResponseDto<Alert>
                {
                    Data = items,
                    CurrentPage = page,
                    PerPage = perPage,
                    Total = total,
                    LastPage = lastPage
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::GetCompanyAlertsAsync failed for company: {CompanyId}", companyId);
                throw;
            }
        }

        public async Task<IEnumerable<Alert>> GetMachineAlertsAsync(long machineId)
        {
            return await _dbContext.Alerts
                .Include(a => a.Acknowledger)
                .Include(a => a.Resolver)
                .Where(a => a.MachineId == machineId)
                .OrderByDescending(a => a.CreatedAt)
                .ToListAsync();
        }

        public async Task<IEnumerable<Alert>> GetCriticalAlertsAsync(long companyId)
        {
            return await _dbContext.Alerts
                .Include(a => a.Machine)
                .Include(a => a.Acknowledger)
                .Include(a => a.Resolver)
                .Where(a => a.CompanyId == companyId && a.Severity == "critical" && (a.Status == "open" || a.Status == "acknowledged"))
                .OrderByDescending(a => a.CreatedAt)
                .ToListAsync();
        }

        public async Task<IEnumerable<AlertRule>> GetAlertRulesAsync(long companyId)
        {
            return await _dbContext.AlertRules
                .Where(r => r.CompanyId == companyId)
                .OrderBy(r => r.Name)
                .ToListAsync();
        }

        public async Task<AlertRule> UpdateAlertRuleAsync(long ruleId, IDictionary<string, object> data)
        {
            try
            {
                var rule = await _dbContext.AlertRules.FindAsync(ruleId);
                if (rule == null)
                {
                    throw new KeyNotFoundException($"Alert rule not found: {ruleId}");
                }

                if (data.TryGetValue("name", out var nameVal)) rule.Name = nameVal.ToString() ?? rule.Name;
                if (data.TryGetValue("is_enabled", out var enabledVal) && enabledVal is bool isEnabled) rule.IsEnabled = isEnabled;
                if (data.TryGetValue("severity", out var sevVal)) rule.Severity = sevVal.ToString() ?? rule.Severity;
                if (data.TryGetValue("threshold_value", out var threshVal) && decimal.TryParse(threshVal.ToString(), out var decVal)) rule.ThresholdValue = decVal;

                await _dbContext.SaveChangesAsync();

                await _auditLogService.LogAsync(
                    EventType.Update.ToString(),
                    $"Alert rule updated: {rule.Name}"
                );

                _logger.LogInformation("Alert rule {RuleId} updated successfully", ruleId);
                return rule;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::UpdateAlertRuleAsync failed for rule ID: {RuleId}", ruleId);
                throw;
            }
        }

        private static bool EvaluateRule(AlertRule rule, MachineCurrentStatus status)
        {
            var metricValue = GetMetricValue(rule.MetricType, status);
            if (metricValue == null) return false;

            var threshold = rule.ThresholdValue;

            return rule.Condition switch
            {
                ">" => metricValue > threshold,
                ">=" => metricValue >= threshold,
                "<" => metricValue < threshold,
                "<=" => metricValue <= threshold,
                "==" => metricValue == threshold,
                "!=" => metricValue != threshold,
                _ => false
            };
        }

        private static decimal? GetMetricValue(string metricType, MachineCurrentStatus status)
        {
            return metricType.ToLowerInvariant() switch
            {
                "cpu_percentage" => status.CpuPercentage,
                "cpu_temperature" => status.CpuTemperature,
                "ram_percentage" => status.RamPercentage,
                "disk_percentage" => status.DiskPercentage,
                "battery_percentage" => status.BatteryPercentage,
                _ => null
            };
        }

        private async Task CreateAlertFromRuleAsync(AlertRule rule, Machine machine, MachineCurrentStatus status)
        {
            try
            {
                var metricValue = GetMetricValue(rule.MetricType, status);
                var alertDescription = $"{rule.MetricType.Replace('_', ' ')} is {rule.Condition} {rule.ThresholdValue} (current: {metricValue})";

                var alertMetadata = new Dictionary<string, object>
                {
                    { "metric", rule.MetricType },
                    { "operator", rule.Condition },
                    { "threshold", rule.ThresholdValue },
                    { "current_value", metricValue ?? 0 }
                };

                var alert = new Alert
                {
                    CompanyId = machine.CompanyId ?? 0,
                    MachineId = machine.Id,
                    AlertRuleId = rule.Id,
                    Severity = rule.Severity,
                    Title = rule.Name,
                    Description = alertDescription,
                    Metadata = JsonSerializer.Serialize(alertMetadata),
                    Status = AlertStatus.Open.ToString().ToLowerInvariant()
                };

                await _dbContext.Alerts.AddAsync(alert);
                await _dbContext.SaveChangesAsync();

                await _auditLogService.LogAsync(
                    EventType.Create.ToString(),
                    $"Alert generated: {rule.Name} for machine: {machine.MachineUid}",
                    machine: machine,
                    newValues: alert
                );

                await _notificationService.SendAlertNotificationAsync(alert);

                if (rule.Severity == "critical" || rule.Severity == "warning")
                {
                    await _notificationService.SendEmailNotificationAsync(alert);
                }

                _logger.LogInformation("Alert created from rule: {AlertId} for machine {MachineId}", alert.Id, machine.Id);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "AlertService::CreateAlertFromRuleAsync failed for rule: {RuleId}", rule.Id);
            }
        }
    }
}
